Python sdk: workflow.py should use use `model_dump(mode="json")` instead of just `model_dump()`

Trying to upgrade to the 1.0+ sdk, but having issues using the new workflow methods.


If you try making a call to `workflow.aio_run_no_wait()` with an input that contains a `datetime` (or another non-jsonifiable member):
```python
workflow_ref = await workflow.aio_run_no_wait(
                        GitHubWebhookWorkflowInput(
                            received_at=received_at,   # datetime.datetime
                            delivery_id=delivery_id,
                            event_type=event_type,
                            request_headers=request_headers,
                            request_body=request_json,
                        ))
```

Hatchet will fail to serialize:

```
ellipsis/src/public_api/routers/github/github_router.py:118: in handle_github_webhook
    workflow_ref = await workflow.aio_run_no_wait(
.venv/lib/python3.11/site-packages/hatchet_sdk/runnables/workflow.py:328: in aio_run_no_wait
    return await self.client._client.admin.aio_run_workflow(
.venv/lib/python3.11/site-packages/tenacity/asyncio/__init__.py:189: in async_wrapped
    return await copy(fn, *args, **kwargs)
.venv/lib/python3.11/site-packages/tenacity/asyncio/__init__.py:111: in __call__
    do = await self.iter(retry_state=retry_state)
.venv/lib/python3.11/site-packages/tenacity/asyncio/__init__.py:153: in iter
    result = await action(retry_state)
.venv/lib/python3.11/site-packages/tenacity/_utils.py:99: in inner
    return call(*args, **kwargs)
.venv/lib/python3.11/site-packages/tenacity/__init__.py:398: in <lambda>
    self._add_action_func(lambda rs: rs.outcome.result())
../../../.pyenv/versions/3.11.6/lib/python3.11/concurrent/futures/_base.py:449: in result
    return self.__get_result()
../../../.pyenv/versions/3.11.6/lib/python3.11/concurrent/futures/_base.py:401: in __get_result
    raise self._exception
.venv/lib/python3.11/site-packages/tenacity/asyncio/__init__.py:114: in __call__
    result = await fn(*args, **kwargs)
.venv/lib/python3.11/site-packages/hatchet_sdk/clients/admin.py:336: in aio_run_workflow
    request = self._create_workflow_run_request(workflow_name, input, options)
.venv/lib/python3.11/site-packages/hatchet_sdk/clients/admin.py:297: in _create_workflow_run_request
    return self._prepare_workflow_request(workflow_name, input, trigger_options)
.venv/lib/python3.11/site-packages/hatchet_sdk/clients/admin.py:105: in _prepare_workflow_request
    payload_data = json.dumps(input)
../../../.pyenv/versions/3.11.6/lib/python3.11/json/__init__.py:231: in dumps
    return _default_encoder.encode(obj)
../../../.pyenv/versions/3.11.6/lib/python3.11/json/encoder.py:200: in encode
    chunks = self.iterencode(o, _one_shot=True)
../../../.pyenv/versions/3.11.6/lib/python3.11/json/encoder.py:258: in iterencode
    return _iterencode(o, 0)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <json.encoder.JSONEncoder object at 0x10e1191d0>, o = datetime.datetime(2025, 4, 10, 1, 20, 24, 737424, tzinfo=datetime.timezone.utc)

    def default(self, o):
        """Implement this method in a subclass such that it returns
        a serializable object for ``o``, or calls the base implementation
        (to raise a ``TypeError``).
    
        For example, to support arbitrary iterators, you could
        implement default like this::
    
            def default(self, o):
                try:
                    iterable = iter(o)
                except TypeError:
                    pass
                else:
                    return list(iterable)
                # Let the base class default method raise the TypeError
                return JSONEncoder.default(self, o)
    
        """
>       raise TypeError(f'Object of type {o.__class__.__name__} '
                        f'is not JSON serializable')
E       TypeError: Object of type datetime is not JSON serializable

../../../.pyenv/versions/3.11.6/lib/python3.11/json/encoder.py:180: TypeError
 ```

Because it's using `model_dump()` instead of `model_dump(mode="json")`. The latter will take into account custom JSON encounters on the target class; e.g. the one defined on the input type i'm using:

```python
class GitHubWebhookWorkflowInput(HasDeliveryId):
    received_at: datetime
    event_type: GitHubWebhookEventType
    request_headers: dict
    request_body: dict

    model_config = pydantic.ConfigDict(
        json_encoders={
            datetime: lambda v: v.isoformat(), # used to serialize effectively
        },
    )
```


This effectively blocks use of the workflows API until it can be fixed.

--- a/sdks/python/hatchet_sdk/runnables/workflow.py
+++ b/sdks/python/hatchet_sdk/runnables/workflow.py
@@ -284,7 +284,7 @@ class BaseWorkflow(Generic[TWorkflowInput]):
     ) -> WorkflowRunTriggerConfig:
         return WorkflowRunTriggerConfig(
             workflow_name=self.config.name,
-            input=input.model_dump() if input else {},
+            input=input.model_dump(mode="json") if input else {},
             options=options,
             key=key,
         )
@@ -303,7 +303,7 @@ class Workflow(BaseWorkflow[TWorkflowInput]):
     ) -> WorkflowRunRef:
         return self.client._client.admin.run_workflow(
             workflow_name=self.config.name,
-            input=input.model_dump() if input else {},
+            input=input.model_dump(mode="json") if input else {},
             options=options,
         )
 
@@ -314,7 +314,7 @@ class Workflow(BaseWorkflow[TWorkflowInput]):
     ) -> dict[str, Any]:
         ref = self.client._client.admin.run_workflow(
             workflow_name=self.config.name,
-            input=input.model_dump() if input else {},
+            input=input.model_dump(mode="json") if input else {},
             options=options,
         )
 
@@ -327,7 +327,7 @@ class Workflow(BaseWorkflow[TWorkflowInput]):
     ) -> WorkflowRunRef:
         return await self.client._client.admin.aio_run_workflow(
             workflow_name=self.config.name,
-            input=input.model_dump() if input else {},
+            input=input.model_dump(mode="json") if input else {},
             options=options,
         )
 
@@ -338,7 +338,7 @@ class Workflow(BaseWorkflow[TWorkflowInput]):
     ) -> dict[str, Any]:
         ref = await self.client._client.admin.aio_run_workflow(
             workflow_name=self.config.name,
-            input=input.model_dump() if input else {},
+            input=input.model_dump(mode="json") if input else {},
             options=options,
         )
 
@@ -389,7 +389,7 @@ class Workflow(BaseWorkflow[TWorkflowInput]):
         return self.client._client.admin.schedule_workflow(
             name=self.config.name,
             schedules=cast(list[datetime | timestamp_pb2.Timestamp], [run_at]),
-            input=input.model_dump() if input else {},
+            input=input.model_dump(mode="json") if input else {},
             options=options,
         )
 
@@ -402,7 +402,7 @@ class Workflow(BaseWorkflow[TWorkflowInput]):
         return await self.client._client.admin.aio_schedule_workflow(
             name=self.config.name,
             schedules=cast(list[datetime | timestamp_pb2.Timestamp], [run_at]),
-            input=input.model_dump(),
+            input=input.model_dump(mode="json"),
             options=options,
         )
 
@@ -417,7 +417,7 @@ class Workflow(BaseWorkflow[TWorkflowInput]):
             workflow_name=self.config.name,
             cron_name=cron_name,
             expression=expression,
-            input=input.model_dump(),
+            input=input.model_dump(mode="json"),
             additional_metadata=additional_metadata,
         )
 
@@ -432,7 +432,7 @@ class Workflow(BaseWorkflow[TWorkflowInput]):
             workflow_name=self.config.name,
             cron_name=cron_name,
             expression=expression,
-            input=input.model_dump(),
+            input=input.model_dump(mode="json"),
             additional_metadata=additional_metadata,
         )
 

--- a/sdks/python/hatchet_sdk/runnables/workflow.py
+++ b/sdks/python/hatchet_sdk/runnables/workflow.py
@@ -284,9 +284,20 @@ class BaseWorkflow(Generic[TWorkflowInput]):
     ) -> WorkflowRunTriggerConfig:
         return WorkflowRunTriggerConfig(
             workflow_name=self.config.name,
-            input=input.model_dump(mode="json") if input else {},
+            input=self._serialize_input(input),
             options=options,
             key=key,
+        )
+
+    def _serialize_input(self, input: TWorkflowInput | None) -> JSONSerializableMapping:
+        if not input:
+            return {}
+
+        if isinstance(input, BaseModel):
+            return input.model_dump(mode="json")
+
+        raise ValueError(
+            f"Input must be a BaseModel or `None`, got {type(input)} instead."
         )
 
 
@@ -303,7 +314,7 @@ class Workflow(BaseWorkflow[TWorkflowInput]):
     ) -> WorkflowRunRef:
         return self.client._client.admin.run_workflow(
             workflow_name=self.config.name,
-            input=input.model_dump(mode="json") if input else {},
+            input=self._serialize_input(input),
             options=options,
         )
 
@@ -314,7 +325,7 @@ class Workflow(BaseWorkflow[TWorkflowInput]):
     ) -> dict[str, Any]:
         ref = self.client._client.admin.run_workflow(
             workflow_name=self.config.name,
-            input=input.model_dump(mode="json") if input else {},
+            input=self._serialize_input(input),
             options=options,
         )
 
@@ -327,7 +338,7 @@ class Workflow(BaseWorkflow[TWorkflowInput]):
     ) -> WorkflowRunRef:
         return await self.client._client.admin.aio_run_workflow(
             workflow_name=self.config.name,
-            input=input.model_dump(mode="json") if input else {},
+            input=self._serialize_input(input),
             options=options,
         )
 
@@ -338,7 +349,7 @@ class Workflow(BaseWorkflow[TWorkflowInput]):
     ) -> dict[str, Any]:
         ref = await self.client._client.admin.aio_run_workflow(
             workflow_name=self.config.name,
-            input=input.model_dump(mode="json") if input else {},
+            input=self._serialize_input(input),
             options=options,
         )
 
@@ -389,20 +400,20 @@ class Workflow(BaseWorkflow[TWorkflowInput]):
         return self.client._client.admin.schedule_workflow(
             name=self.config.name,
             schedules=cast(list[datetime | timestamp_pb2.Timestamp], [run_at]),
-            input=input.model_dump(mode="json") if input else {},
+            input=self._serialize_input(input),
             options=options,
         )
 
     async def aio_schedule(
         self,
         run_at: datetime,
-        input: TWorkflowInput,
+        input: TWorkflowInput | None = None,
         options: ScheduleTriggerWorkflowOptions = ScheduleTriggerWorkflowOptions(),
     ) -> WorkflowVersion:
         return await self.client._client.admin.aio_schedule_workflow(
             name=self.config.name,
             schedules=cast(list[datetime | timestamp_pb2.Timestamp], [run_at]),
-            input=input.model_dump(mode="json"),
+            input=self._serialize_input(input),
             options=options,
         )
 
@@ -410,14 +421,14 @@ class Workflow(BaseWorkflow[TWorkflowInput]):
         self,
         cron_name: str,
         expression: str,
-        input: TWorkflowInput,
-        additional_metadata: JSONSerializableMapping,
+        input: TWorkflowInput | None = None,
+        additional_metadata: JSONSerializableMapping = {},
     ) -> CronWorkflows:
         return self.client.cron.create(
             workflow_name=self.config.name,
             cron_name=cron_name,
             expression=expression,
-            input=input.model_dump(mode="json"),
+            input=self._serialize_input(input),
             additional_metadata=additional_metadata,
         )
 
@@ -425,14 +436,14 @@ class Workflow(BaseWorkflow[TWorkflowInput]):
         self,
         cron_name: str,
         expression: str,
-        input: TWorkflowInput,
-        additional_metadata: JSONSerializableMapping,
+        input: TWorkflowInput | None = None,
+        additional_metadata: JSONSerializableMapping = {},
     ) -> CronWorkflows:
         return await self.client.cron.aio_create(
             workflow_name=self.config.name,
             cron_name=cron_name,
             expression=expression,
-            input=input.model_dump(mode="json"),
+            input=self._serialize_input(input),
             additional_metadata=additional_metadata,
         )
 

--- a/sdks/python/pyproject.toml
+++ b/sdks/python/pyproject.toml
@@ -1,6 +1,6 @@
 [tool.poetry]
 name = "hatchet-sdk"
-version = "1.2.5"
+version = "1.2.6"
 description = ""
 authors = ["Alexander Belanger <alexander@hatchet.run>"]
 readme = "README.md"

